ui-tests/tests/widget_notebook_example.test.ts

```diff
@@ -0,0 +1,139 @@
+      path.resolve(__dirname, "../../demo-notebooks/guided-demos"),
+      tmpPath
+    );
+    await page.filebrowser.openDirectory(tmpPath);
+  });
+
+  test("Run notebook, capture cell outputs, and test widgets", async ({
+    page,
+    tmpPath,
+  }) => {
+    const notebook = "3_widget_example.ipynb";
+    await page.notebook.openByPath(`${tmpPath}/${notebook}`);
+    await page.notebook.activate(notebook);
+
+    const captures: (Buffer | null)[] = []; // Array to store cell screenshots
+    const cellCount = await page.notebook.getCellCount();
+
+    // Run all cells and capture their screenshots
+    await page.notebook.runCellByCell({
+      onAfterCellRun: async (cellIndex: number) => {
+        const cell = await page.notebook.getCellOutput(cellIndex);
+        if (cell && (await cell.isVisible())) {
+          captures[cellIndex] = await cell.screenshot(); // Save the screenshot by cell index
+        }
+        await page.addStyleTag({ content: '.jp-cell-toolbar { display: none !important; }' });
+      },
+    });
+
+    await page.notebook.save();
+
+    // Ensure that each cell's screenshot is captured
+    for (let i = 0; i < cellCount; i++) {
+      const image = `widgets-cell-${i}.png`;
+
+      if (captures[i]) {
+        expect.soft(captures[i]).toMatchSnapshot(image); // Compare pre-existing capture
+        continue;
+      }
+    }
+
+    const widgetCellIndex = 3;
+
+    await waitForWidget(page, widgetCellIndex, 'input[type="checkbox"]');
+    await waitForWidget(page, widgetCellIndex, 'button:has-text("Cluster Down")');
+    await waitForWidget(page, widgetCellIndex, 'button:has-text("Cluster Up")');
+
+    await interactWithWidget(page, widgetCellIndex, 'input[type="checkbox"]', async (checkbox) => {
+      await checkbox.click();
+      const isChecked = await checkbox.isChecked();
+      expect(isChecked).toBe(true);
+    });
+
+    await interactWithWidget(page, widgetCellIndex, 'button:has-text("Cluster Down")', async (button) => {
+      await button.click();
+      const clusterDownMessage = await page.waitForSelector('text=No instances found, nothing to be done.', { timeout: 5000 });
+      expect(clusterDownMessage).not.toBeNull();
+    });
+
+    await interactWithWidget(page, widgetCellIndex, 'button:has-text("Cluster Up")', async (button) => {
+      await button.click();
+
+      const successMessage = await page.waitForSelector('text=Ray Cluster: \'raytest\' has successfully been created', { timeout: 10000 });
+      expect(successMessage).not.toBeNull();
+
+      const resourcesMessage = await page.waitForSelector('text=Waiting for requested resources to be set up...');
+      expect(resourcesMessage).not.toBeNull();
+
+      const upAndRunningMessage = await page.waitForSelector('text=Requested cluster is up and running!');
+      expect(upAndRunningMessage).not.toBeNull();
+
+      const dashboardReadyMessage = await page.waitForSelector('text=Dashboard is ready!');
```

Comment:
Should we use timeouts for all of these just incase the Dashboard doesn't come up (or someone changes the text but forgets to change the test)?

Reply:
It will still wait for the timeout of the set time out

Reply:
Yeah, that's fine. I just mean so that if the changes are pushed the test doesn't hang indefinitely waiting for the condition

Reply:
But if it wait's for the cluster to be ready and it can take sometime or even fail, what is the timeout for that? The cluster is ready at different time every single time